common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/QueryContentProviderLoader.kt

```diff
@@ -0,0 +1,102 @@
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.broker.ipc
+
+import android.content.Context
+import android.content.pm.PackageManager
+import android.content.pm.ProviderInfo
+import android.os.Build
+import android.os.DeadObjectException
+import com.microsoft.identity.common.java.interfaces.INameValueStorage
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents
+import com.microsoft.identity.common.logging.Logger
+
+/**
+ * packageManager.queryContentProviders involves an ipc call (we've seen DeadObjectException from it before)
+ *
+ * We'll avoid making unnecessary ipc request by caching them.
+ * The content provider support status should remain the same for a given broker version.
+ **/
+class QueryContentProviderLoader(
+    private val getVersionCode: (appPackageName: String) -> String,
+    private val supportedByContentProvider: (appPackageName: String) -> Boolean,
+    private val fileManager: INameValueStorage<String>,
+) : IQueryContentProviderLoader {
+
+    companion object {
+        private val TAG = QueryContentProviderLoader::class.java.simpleName
+        private const val SHARED_PREFERENCE_NAME = "com.microsoft.common.ipc.content.provider.query.cache"
+
+        @Throws(PackageManager.NameNotFoundException::class)
+        private fun getVersionCode(context: Context, appPackageName: String) : String {
+            val packageInfo = context.packageManager.getPackageInfo(appPackageName, 0)
+            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                packageInfo.longVersionCode.toString()
+            } else {
+                packageInfo.versionCode.toString()
+            }
+        }
+
+        @Throws(DeadObjectException::class)
+        fun supportedByContentProvider(context: Context, appPackageName: String) : Boolean {
+            val contentProviderAuthority: String =
+                ContentProviderStrategy.getContentProviderAuthority(appPackageName)
+
+            val providers: List<ProviderInfo> = context.packageManager
+                .queryContentProviders(null, 0, 0)
+
+            for (providerInfo in providers) {
+                if (providerInfo.authority != null && providerInfo.authority == contentProviderAuthority) {
+                    return true
+                }
+            }
+
+            return false
+        }
+    }
+
+    constructor(context: Context, components: IPlatformComponents): this(
+        fileManager = components.storageSupplier.getUnencryptedNameValueStore(
+            SHARED_PREFERENCE_NAME, String::class.java),
+        getVersionCode = { pkgName -> getVersionCode(context, pkgName) },
+        supportedByContentProvider = { pkgName -> supportedByContentProvider(context, pkgName) }
+    )
+
+    override fun getResult(packageName: String) : Boolean {
+        val methodTag = "$TAG:getResult"
+        try {
+            val key = packageName + ":" + getVersionCode(packageName)
```

Comment:
Can you put some comments here to describe what's we are doing. It looks like we are using the concatenation of package name and package version to create a cache key for caching whether that given package/version pair supports content provider or not, but I think comments should make it more clear. Overall, I think this class can benefit from some comments. 

Also not a huge fan of `getResult` method name. And I'm also wondering if version code should also become a param to this method. So the caller has to supply both package name and version code to avoid any assumptions about what the method does internally. Alternatively sufficient javadoc can also help avoid these confusions.

Reply:
Version code is an internal implementation of this class.

This class is a wrapper over context.packageManager.queryContentProviders, and should take only the package name of the targeted app. caching logic (how key is constructed, how data is persisted) is internal.